libdevcrypto/gm/sm2/sm2.cpp

```diff
@@ -21,353 +21,199 @@
+      CRYPTO_LOG(ERROR) << "[SM2::veify] ERROR of Verify EC_POINT_hex2point"
+            ;
+      goto done;
     }
 
-    if (!EC_POINT_hex2point(sm2Group, (const char*)publicKey.c_str(), pubPoint, NULL))
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::veify] ERROR of Verify EC_POINT_hex2point"
-                          << LOG_KV("pubKey", publicKey);
-        goto err;
+    sm2Key = EC_KEY_new();
+    if (sm2Key == NULL) {
+      CRYPTO_LOG(ERROR) << "[SM2::verify] ERROR of EC_KEY_new";
+      goto done;
     }
 
-    sm2Key = EC_KEY_new_by_curve_name(NID_sm2);
-
-    if (sm2Key == NULL)
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::veify] ERROR of Verify EC_KEY_new_by_curve_name"
-                          << LOG_KV("pubKey", publicKey);
-        goto err;
+    if (!EC_KEY_set_group(sm2Key, sm2Group)) {
+      CRYPTO_LOG(ERROR) << "[SM2::verify] ERROR of EC_KEY_set_group";
+      goto done;
     }
-
-    if (!EC_KEY_set_public_key(sm2Key, pubPoint))
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::veify] ERROR of Verify EC_KEY_set_public_key"
-                          << LOG_KV("pubKey", publicKey);
-        goto err;
+    if (!EC_KEY_set_public_key(sm2Key, point)) {
+      CRYPTO_LOG(ERROR) << "[SM2::verify] ERROR of EC_KEY_set_public_key";
+      goto done;
     }
-
-    if (!ECDSA_sm2_get_Z((const EC_KEY*)sm2Key, NULL, NULL, 0, zValue, &zValueLen))
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::veify] Error Of Compute Z" << LOG_KV("pubKey", publicKey);
-        goto err;
+    sig = ECDSA_SIG_new();
+    if (!BN_hex2bn(&rBigNum, r.c_str())) {
+      CRYPTO_LOG(ERROR) << "[SM2::verify] ERROR of BN_hex2bn for r:" << r;
+      goto done;
     }
-    // SM3 Degist
-    SM3_Init(&sm3Ctx);
-    SM3_Update(&sm3Ctx, zValue, zValueLen);
-    SM3_Update(&sm3Ctx, originalData, originalDataLen);
-    SM3_Final(zValue, &sm3Ctx);
 
-    /*Now Verify it*/
-    signData = ECDSA_SIG_new();
-    if (!BN_hex2bn(&signData->r, r.c_str()))
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::veify] ERROR of BN_hex2bn" << LOG_KV("R", r)
-                          << LOG_KV("pubKey", publicKey);
-        goto err;
+    if (!BN_hex2bn(&sBigNum, s.c_str())) {
+      CRYPTO_LOG(ERROR) << "[SM2::verify] ERROR of BN_hex2bn for s:" << s;
+      goto done;
     }
-
-    if (!BN_hex2bn(&signData->s, s.c_str()))
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::veify] ERROR BN_hex2bn:" << LOG_KV("S", s)
-                          << LOG_KV("pubKey", publicKey);
-        goto err;
+    if (!ECDSA_SIG_set0(sig, rBigNum, sBigNum)) {
+      CRYPTO_LOG(ERROR) << "[SM2::verify] ERROR of ECDSA_SIG_set0 failed";
+      goto done;
     }
-
-    if (ECDSA_do_verify(zValue, zValueLen, signData, sm2Key) != 1)
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::veify] Error Of SM2 Verify" << LOG_KV("pubKey", publicKey);
-        goto err;
+    ok = sm2_do_verify(sm2Key, EVP_sm3(), sig, (const uint8_t *)userId,
+                       strlen(userId), (const uint8_t *)_originalData,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/gh/FISCO-BCOS/FISCO-BCOS/pullRequest?prid=5347884)